R/flatten.R

```diff
@@ -0,0 +1,118 @@
+#' Flatten `dm` into a wide table
+#'
+#' This function joins all the tables of your `dm` object together and returns one
+#' table containing all unique columns. This is especially useful in connection
+#' with `cdm_select_tbl()` in order to reduce the number of tables in advance.
+#'
+#' @inheritParams cdm_join_tbl
+#' @param sep In case there are columns with non-unique names, they will be renamed using
+#' [`cdm_disambiguate_cols()`]. `sep` is the character variable separating the table
+#' names and the ambiguous column names
+#' @family Flattening functions
+#'
+#' @return A wide table resulting of consecutively joining all tables together.
+#'
+#' @examples
+#' cdm_nycflights13() %>%
+#' cdm_select_tbl(-weather) %>%
+#' cdm_flatten()
+#'
+#' @export
+cdm_flatten <- function(dm, join = left_join, sep = ".") {
+  if (!are_all_vertices_connected(create_graph_from_dm(dm), src_tbls(dm))) {
+    abort_vertices_not_connected("cdm_flatten")
+  }
+  clean_dm <- cdm_apply_filters(dm) %>%
+      cdm_disambiguate_cols(sep = sep, quiet = TRUE) %>%
+      adapt_fk_cols()
+  g <- create_graph_from_dm(clean_dm, directed = TRUE)
+  filtered_tables <- clean_dm %>% cdm_get_tables()
+
+  # chose a "child table" (only outgoing FKs) as a start (otherwise we might
+  # end up with empty rows, if "parent/dim table" contains PK values that are
+  # not present in child table (we might anyway though, depending on `join`-type))
+  initial_LHS_name <- g %>%
```

Comment:
`names(igraph::topo_sort(g))[[1]]` -- let's reserve the pipe for data frame manipulations.

(What do we do with a `dm` with zero tables?)

Reply:
Throw an error if trying to flatten an empty list of tables? we could call in the beginning `is_empty(cdm_get_tables(dm))`. Or maybe better: 
```r
list_of_tables <- cdm_get_tables(dm)
if (is_empty(cdm_get_tables(dm))) ...
if (nrow(list_of_tables == 1)) return(...)
```
Then we'd have all (or most) corner cases sorted

Reply:
A 0x0 tibble seems like the logical result. Do we really need to special-case `nrow() == 1` ?

Reply:
This corner case no longer occurs with the current interface.